doc/developer/scan.txt

```diff
@@ -132,22 +63,266 @@ files, and section of the code they deal with, are :
+on small tensors by few operations on large tensors. It can also lead to
+increased memory usage.
+
+
+PushOutDot1
+-----------
+
+This is another optimization that attempts to detect certain patterns of
+computation in a Scan op's inner function and move this computation to the
+outer graph.
+
+
+ScanInplaceOptimizer
+--------------------
+
+This optimization attempts to make Scan compute its recurrent outputs inplace
+on the input tensors that contain their initial states. This optimization can
+improve runtime performance as well as reduce memory usage.
+
+
+ScanSaveMem
+-----------
+
+This optimizations attempts to determine if a Scan node, during its execution,
+for any of its outputs, can get away with allocating a memory buffer that is
+large enough to contain some of the computed timesteps of that output but not
+all of them.
+
+By default, during the execution of a Scan node, memory buffers will be
+allocated to store the values computed for every output at every iteration.
+However, in some cases, there are outputs for which there is only really a
+need to store the most recent N values, not all of them.
+
+For instance, if a Scan node has a sitsot output (last computed value is
+fed back as an input at the next iteration) and only the last timestep of
+that output is ever used in the outer function, the ScanSaveMem optimization
+could determine that there is no need to store all computed timesteps for
+that sitsot output. Only the most recently computed timestep ever needs to
+be kept in memory.
+
+
+ScanMerge
+---------
+
+This optimization attempts to fuse distinct Scan ops into a single Scan op
+that performs all the computation. The main advantage of merging Scan ops
+together comes from the possibility of both original ops having some
+computation in common. In such a setting, this computation ends up being done
+twice. The fused Scan op, however, would only need to do it once and could
+therefore be more computationally efficient. Also, since every Scan node
+involves a certain overhead, at runtime, reducing the number of Scan nodes in
+the graph can improve performance.
+
+
+scan_merge_inouts
+-----------------
+
+This optimization attempts to merge a Scan op's identical outer inputs as well
+as merge its identical outer outputs (outputs that perform the same
+computation on the same inputs). This can reduce the amount of computation as
+well as result in a simpler graph for both the inner function and the outer
+function.
+
+
+Helper classes and functions
+============================
+
+Because of the complexity involved in dealing with Scan, a large number of
+helper classes and functions have been developped over time to implement
+operations commonly needed when dealing with the Scan op. The Scan op
+itself defines a large number of them and others can be found in the file
+`scan_utils.py`. This sections aims to point out the most useful ones and
+organize them according to their uses.
```

Comment:
Maybe

"ones and organize them according to their uses."
->
"ones sorted by usage."

If that's what you meant.